### PR TITLE
Extend API logging 

### PR DESCRIPTION
**Pull request by ewelinagr**
As a system administrator, I want to be able to see a log of relevant calls to the REST API (TMT-181):
TMT-180 - Add audit interceptors to log data requests for auditing
TMT-251 - Add saving API call logs to the file, fix an audit logger configuration

### CODE REVIEW

**Comment by ewelinagr** [comment]
Audit logger is enabled  by default here for both transmart-server and transmart-api-server, maybe this is not required, so it could be changed after this functionality will be tested.

File: transmart-api-server/grails-app/conf/logback.groovy [hunk 1]
```diff
@@ -37,3 +40,47 @@ if (Environment.isDevelopmentMode() && targetDir != null) {
...
+        // daily rollover
+        fileNamePattern = "${logDirectory}/audit.%d{yyyy-MM-dd}.%i.log"
+        // size limit on the log files
+        maxFileSize = '100MB'
+        /* // optional parameters
+        // maxHistory controls the maximum number of archive files to keep, asynchronously deleting older files.
+        maxHistory = 30
+        // totalSizeCap controls the total size of all archive files.
+        // Oldest archives are deleted asynchronously when the total size cap is exceeded
+        totalSizeCap = '3GB'
```
Comment (ewelinagr) [inline transmart-api-server/grails-app/conf/logback.groovy]: this parameters are also useful, but are optional, so I commented them out for now. Please let me know, if it is better to have them there by default.

File: transmart-api-server/grails-app/conf/logback.groovy [hunk 2]
```diff
@@ -37,3 +40,47 @@ if (Environment.isDevelopmentMode() && targetDir != null) {
...
+        }
+    }
+}
+
+appender('processAuditLogger', ChildProcessAppender) {
+    command = ['/usr/bin/your/command/here', 'arg1', 'arg2']
+}
+
+logger('org.transmartproject.db.log', TRACE, ['fileAuditLogger'], true)
+logger('org.transmartproject.db.log', TRACE, ['processAuditLogger'], true)
```
Comment (ewelinagr) [inline transmart-api-server/grails-app/conf/logback.groovy]: I limited the audit logs here only to API ones (they are added in `org.transmartproject.db.log.AccessLogService.groovy`), so let me know if you think we want to have also other logs in audit.log file.

**Comment by ewelinagr** [comment]
Entries in audit.log file look like this now: 
`{
  "timestamp" : "1529410978614",
  "level" : "TRACE",
  "thread" : "http-nio-8080-exec-5",
  "logger" : "org.transmartproject.db.log.AccessLogService",
  "message" : "{\"username\":\"test-public-user-1\",\"event\":\"Concepts retrieval\",\"eventMessage\":\"User (IP: 127.0.0.1) made a concept request.\",\"requestURL\":\"/v2/concepts\",\"accessTime\":\"2018-06-19T12:22:58+0000\"}",
  "context" : "default"
}`

So my question is if this is fine or we want to have only the message part in there?

File: transmart-api-server/grails-app/conf/logback.groovy [hunk 3]
```diff
@@ -37,3 +40,47 @@ if (Environment.isDevelopmentMode() && targetDir != null) {
...
+            jsonFormatter(JacksonJsonFormatter) {
+                prettyPrint = true
+            }
+            appendLineSeparator = true
+        }
+    }
+}
+
+appender('processAuditLogger', ChildProcessAppender) {
+    command = ['/usr/bin/your/command/here', 'arg1', 'arg2']
```
Comment (gijskant) [inline transmart-api-server/grails-app/conf/logback.groovy]: Won't this fail if not commented out by default?

File: transmart-core-db/grails-app/services/org/transmartproject/db/log/AccessLogService.groovy [hunk 4]
```diff
@@ -3,14 +3,25 @@
...
 import org.transmartproject.core.users.User
 
 @Transactional
 class AccessLogService implements AccessLogEntryResource {
 
+    static final Logger log = LoggerFactory.getLogger(AccessLogService.class)
+
     @Override
     AccessLogEntry report(Map<String, Object> additionalParams = [:], User user, String event) {
+        log.trace(new JsonBuilder([
```
Comment (gijskant) [inline transmart-core-db/grails-app/services/org/transmartproject/db/log/AccessLogService.groovy]: In other places, we try to stick to Jackson for (de)serialising JSON, e.g., 
```java
new ObjectMapper.writeValueAsString(value)
```
(or GSON for streaming serialisation... and trying to get of the Grails specific JSON converters...)

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/AuditInterceptor.groovy [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package org.transmartproject.interceptors
+
+import groovy.transform.CompileStatic
+import org.transmartproject.core.log.AccessLogEntryResource
+import org.transmartproject.core.users.User
+import org.transmartproject.rest.user.AuthContext
+
+@CompileStatic
+class AuditInterceptor {
```
Comment (gijskant) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/AuditInterceptor.groovy]: Why not make this abstract?

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy [hunk 6]
```diff
@@ -0,0 +1,57 @@
+package org.transmartproject.interceptors
+
+import groovy.transform.CompileStatic
+
+@CompileStatic
+class QueryAuditInterceptor extends AuditInterceptor {
```
Comment (gijskant) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy]: This class looks really high-maintainance...

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/DimensionAuditInterceptor.groovy [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package org.transmartproject.interceptors
+
+import groovy.transform.CompileStatic
+
+@CompileStatic
+class DimensionAuditInterceptor extends AuditInterceptor {
```
Comment (gijskant) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/DimensionAuditInterceptor.groovy]: A separate audit interceptor class for every controller... isn't that a bit too much?

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                return [event       : 'Observations cross table retrieval',
+                        eventMessage: "User (IP: ${ip}) made a cross table request."] as Map<String, String>
+            case 'table':
+                return [event       : 'Observations tabular view retrieval',
+                        eventMessage: "User (IP: ${ip}) made a data table request."] as Map<String, String>
+            case 'supportedFields':
+                return [event       : 'Supported fields view retrieval',
+                        eventMessage: "User (IP: ${ip}) made a supported fields request."] as Map<String, String>
+        }
+        [event: '', eventMessage: ''] as Map<String, String>
```
Comment (gijskant) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy]: Empty default?

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/SystemAuditInterceptor.groovy [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+@CompileStatic
+class SystemAuditInterceptor extends AuditInterceptor {
+
+    SystemAuditInterceptor() {
+        match(controller: ~/system/, action: ~/afterDataLoadingUpdate/)
+    }
+
+    boolean after() {
+        def ip = getIP()
```
Comment (gijskant) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/SystemAuditInterceptor.groovy]: This line is not needed, right? Groovy will try to call `getIp()` if you use `${ip}`.

File: transmartApp/grails-app/conf/logback.groovy [hunk 10]
```diff
@@ -50,34 +53,43 @@ if (productionMode && logDirectory) {
...
+
 /**
  * Configuration for writing audit metrics.
  * This needs to be placed in the out-of-tree Config.groovy, as the log4j config there will override this.
- * (and don't forget to 'import org.apache.log4j.DailyRollingFileAppender',
- * 'import org.transmart.server.logging.ChildProcessAppender' and 'import org.transmart.server.logging.JsonLayout'.)
+ * See https://logback.qos.ch/manual/appenders.html for details on configuration
  */
 /*
+appender('fileAuditLogger', RollingFileAppender) {
```
Comment (gijskant) [inline transmartApp/grails-app/conf/logback.groovy]: Why is this stuff in the transmartApp logback config, can that file not be removed completely?

File: transmart-core-db/grails-app/services/org/transmartproject/db/log/AccessLogService.groovy [hunk 11]
```diff
@@ -3,14 +3,25 @@
...
 import org.transmartproject.core.log.AccessLogEntryResource
 import org.transmartproject.core.users.User
 
 @Transactional
 class AccessLogService implements AccessLogEntryResource {
 
+    static final Logger log = LoggerFactory.getLogger(AccessLogService.class)
+
     @Override
     AccessLogEntry report(Map<String, Object> additionalParams = [:], User user, String event) {
```
Comment (gijskant) [inline transmart-core-db/grails-app/services/org/transmartproject/db/log/AccessLogService.groovy]: These logs will also end up in the database (and if transmartApp is available, in the admin panel there). Seems like these things should not be mixed.

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+                action: ~/observations|counts|countsPerConcept|countsPerStudy|countsPerStudyAndConcept|aggregatesPerConcept|crosstable|table|supportedFields/)
+    }
+
+    boolean after() {
+        def ip = getIP()
+        Map<String, String> event = getEventMessage(ip)
+        report(event.event, event.eventMessage)
+    }
+
+    private Map<String, String> getEventMessage(ip) {
```
Comment (gijskant) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy]: They are really nice descriptive messages, but the audit logs are mainly meant for processing in other systems, not to be readable in a UI, so this seems a bit overkill.

File: transmart-core-db/grails-app/services/org/transmartproject/db/log/AccessLogService.groovy (hunk 11 above)
Comment (ewelinagr) [reply transmart-core-db/grails-app/services/org/transmartproject/db/log/AccessLogService.groovy]: But the requirement is to add them both to the database and (if configured) to the log file. You mean it should be a separate function or I should use a separate services?

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/DimensionAuditInterceptor.groovy (hunk 7 above)
Comment (ewelinagr) [reply transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/DimensionAuditInterceptor.groovy]: hmm, ok, I wanted them to be more descriptive, but if it is not needed, I can move everything to one interceptor that will log information containing message with a controller name and action. Would it be sufficient?

Comment (ewelinagr) [reply transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/DimensionAuditInterceptor.groovy]: The only disadvantage is that an action name is not always describing the action well, but you mentioned they do not have to be readable in UI, so should be fine.

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy (hunk 6 above)
Comment (ewelinagr) [reply transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy]: I reduced the message generation part, now it's as other interceptors

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy [hunk 13]
```diff
@@ -0,0 +1,17 @@
+package org.transmartproject.interceptors
+
+import groovy.transform.CompileStatic
+
+@CompileStatic
+class QueryAuditInterceptor extends AuditInterceptor {
+
+    QueryAuditInterceptor() {
+        match(controller: ~/query/,
+                action: ~/observations|counts|countsPerConcept|countsPerStudy|countsPerStudyAndConcept|aggregatesPerConcept|crosstable|table|supportedFields/)
```
Comment (gijskant) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/QueryAuditInterceptor.groovy]: You could instead blacklist calls that should not be logged. Imagine the case when you extend the query controller, do we then expect that this interceptor is updated as well?

File: transmart-rest-api/src/main/groovy/org/transmartproject/rest/logging/ChildProcessAppender.groovy [hunk 14]
```diff
@@ -79,29 +81,32 @@ class ChildProcessAppender extends AppenderSkeleton {
     protected boolean broken = false
 
     private ThreadLocal<int[]> recursionCount = [
-        initialValue: { [0] as int[] }
+            initialValue: { [0] as int[] }
     ] as ThreadLocal<int[]>
 
     void setRestartLimit(int l) {
-        if (l < 0) throw new IllegalArgumentException("restartLimit cannot be negative (use 0 to disable the limit)")
+        if (l < 0)
```
Comment (gijskant) [inline transmart-rest-api/src/main/groovy/org/transmartproject/rest/logging/ChildProcessAppender.groovy]: I think we aggreed on always using
```java
if (condition) {
  // body
}
```
with the `{}` around the body.

File: transmartApp/grails-app/controllers/org/transmartproject/interceptors/AuditLogService.groovy [hunk 15]
```diff
@@ -17,7 +17,7 @@
  * Transmart.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-package org.transmart.server
+package org.transmartproject.interceptors
```
Comment (gijskant) [inline transmartApp/grails-app/controllers/org/transmartproject/interceptors/AuditLogService.groovy]: But this class is not an interceptor, right?

**Comment by ewelinagr** [comment]
@gijskant Please check my latest commit. Log entries look like this now: 
```
{
  "username" : "01ec82a4-dd7d-45ff-ac14-c5e10c20eb44",
  "event" : "query request",
  "eventMessage" : "{\"ip\":\"127.0.0.1\",\"action\":\"table\",\"body\":\"[columnSort:[[sortOrder:asc, dimension:trial visit], [sortOrder:desc, dimension:concept]], offset:2, rowDimensions:[patient, study], columnDimensions:[trial visit, concept], limit:2, constraint:[conceptCode:EHR:VSIGN:HR, type:concept], type:clinical, rowSort:[[sortOrder:desc, dimension:patient]]]\"}",
  "requestURL" : "/v2/observations/table",
  "accessTime" : "2018-07-04T15:31:35"
}
```

File: transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/ApiAuditInterceptor.groovy [hunk 16]
```diff
@@ -33,30 +33,37 @@ class ApiAuditInterceptor {
...
+        return "${request.forwardURI}${request.queryString ? '?' + request.queryString : ''}"
+    }
+
+    protected String getEventMessage() {
+        Map<String, String> message = [
+                ip    : IP,
+                action: actionName
+        ]
+        if ("POST".equalsIgnoreCase(request.getMethod())) {
+            message.put("body", request.JSON.toString())
```
Comment (gijskant) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/interceptors/ApiAuditInterceptor.groovy]: Instead you could also create message as `Map<String, Object>` and use `message.put("body", request.JSON as Map)`.

**Pull request merged**